File: frontend/src/mocks/supplier.ts
import type { Supplier } from '../types/supplier';

/** モック取引先データ */
export const mockSuppliers: Supplier[] = [
  {
    id: 1,
    name: '〇〇農園',
    description:
      '富山県で無農薬野菜を栽培されています。旬の野菜を直接仕入れています。',
    instagram_url: 'https://www.instagram.com/example_farm/',
    image_url: null,
    display_order: 1,
    is_active: true,
    created_at: '2026-01-15T10:00:00+09:00',
    updated_at: '2026-01-15T10:00:00+09:00',
  },
  {
    id: 2,
    name: '△△和紅茶園',
    description:
      '国産和紅茶を生産されています。当店の和紅茶はこちらから仕入れています。',
    instagram_url: 'https://www.instagram.com/example_tea/',
    image_url: null,
    display_order: 2,
    is_active: true,
    created_at: '2026-01-15T10:00:00+09:00',
    updated_at: '2026-01-15T10:00:00+09:00',
  },
  {
    id: 3,
    name: '□□水産',
    description:
      '富山湾の新鮮な魚を毎朝届けていただいています。',
    instagram_url: 'https://www.instagram.com/example_fish/',
    image_url: null,
    display_order: 3,
    is_active: false,
    created_at: '2026-01-20T10:00:00+09:00',
    updated_at: '2026-02-01T10:00:00+09:00',
  },
  {
    id: 4,
    name: '◎◎養蜂場',
    description:
      '富山県産の天然はちみつを生産されています。季節のデザートに使用しています。',
    instagram_url: 'https://www.instagram.com/example_honey/',
    image_url: null,
    display_order: 4,
    is_active: true,
    created_at: '2026-02-10T10:00:00+09:00',
    updated_at: '2026-02-10T10:00:00+09:00',
  },
];

/**
 * 取引先一覧を取得（モック）
 * TODO: GET /api/v1/admin/suppliers に置き換え
 */
export function getSuppliers(activeOnly = false): Supplier[] {
  let suppliers = [...mockSuppliers].sort(
    (a, b) => a.display_order - b.display_order,
  );
  if (activeOnly) {
    suppliers = suppliers.filter((s) => s.is_active);
  }
  return suppliers;
}

/**
 * 取引先を取得（モック）
 * TODO: GET /api/v1/admin/suppliers/:id に置き換え
 */
export function getSupplier(id: number): Supplier | undefined {
  return mockSuppliers.find((s) => s.id === id);
}

File: backend/internal/handler/supplier.go
package handler

import (
	"errors"
	"log"
	"net/http"

	usecase "github.com/cergijame101007/satehits/internal/application/usecase/supplier"
	"github.com/cergijame101007/satehits/internal/domain"
)

// PublicSupplierResponse は公開取引先レスポンス DTO（OpenAPI PublicSupplier）
type PublicSupplierResponse struct {
	ID           int64   `json:"id"`
	Name         string  `json:"name"`
	Description  string  `json:"description"`
	InstagramURL *string `json:"instagram_url"`
	ImageURL     *string `json:"image_url"`
}

// PublicSupplierListResponse は公開取引先一覧レスポンス DTO（OpenAPI SupplierListResponse）
type PublicSupplierListResponse struct {
	Suppliers []PublicSupplierResponse `json:"suppliers"`
}

// SupplierResponse は管理者向け取引先レスポンス DTO（OpenAPI SupplierResponse）
type SupplierResponse struct {
	ID           int64   `json:"id"`
	Name         string  `json:"name"`
	Description  string  `json:"description"`
	InstagramURL *string `json:"instagram_url"`
	ImageURL     *string `json:"image_url"`
	DisplayOrder int     `json:"display_order"`
	IsActive     bool    `json:"is_active"`
	CreatedAt    string  `json:"created_at"`
	UpdatedAt    string  `json:"updated_at"`
}

// AdminSupplierListResponse は管理者向け取引先一覧レスポンス DTO（OpenAPI AdminSupplierListResponse）
type AdminSupplierListResponse struct {
	Suppliers []SupplierResponse `json:"suppliers"`
}

// PublicSupplierHandler は公開取引先一覧 HTTP ハンドラ
type PublicSupplierHandler struct {
	listSuppliers       *usecase.ListSuppliersUseCase
	publicSuppliersPath string
}

// NewPublicSupplierHandler は PublicSupplierHandler を生成する
func NewPublicSupplierHandler(listSuppliers *usecase.ListSuppliersUseCase, publicSuppliersPath string) *PublicSupplierHandler {
	return &PublicSupplierHandler{
		listSuppliers:       listSuppliers,
		publicSuppliersPath: publicSuppliersPath,
	}
}

// HandlePublicSuppliers は GET /api/v1/suppliers をルーティングする（is_active=true のみ）
func (h *PublicSupplierHandler) HandlePublicSuppliers(w http.ResponseWriter, r *http.Request) {
	if r.URL.Path != h.publicSuppliersPath {
		http.NotFound(w, r)
		return
	}
	if r.Method != http.MethodGet {
		http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
		return
	}

	// 公開一覧は常に activeOnly=true 固定。クエリパラメータでの絞り込みは受け付けない。
	suppliers, err := h.listSuppliers.Execute(r.Context(), true)
	if writeSupplierUsecaseError(w, err, "Failed to list public suppliers") {
		return
	}

	responses := make([]PublicSupplierResponse, len(suppliers))
	for i, s := range suppliers {
		responses[i] = toPublicSupplierResponse(s)
	}
	respondWithJSON(w, http.StatusOK, PublicSupplierListResponse{Suppliers: responses})
}

func toPublicSupplierResponse(s domain.Supplier) PublicSupplierResponse {
	return PublicSupplierResponse{
		ID:           s.ID,
		Name:         s.Name,
		Description:  s.Description,
		InstagramURL: s.InstagramURL,
		ImageURL:     s.ImageURL,
	}
}

func toSupplierResponse(s domain.Supplier) SupplierResponse {
	return SupplierResponse{
		ID:           s.ID,
		Name:         s.Name,
		Description:  s.Description,
		InstagramURL: s.InstagramURL,
		ImageURL:     s.ImageURL,
		DisplayOrder: s.DisplayOrder,
		IsActive:     s.IsActive,
		CreatedAt:    s.CreatedAt.Format(timeRFC3339),
		UpdatedAt:    s.UpdatedAt.Format(timeRFC3339),
	}
}

// writeSupplierUsecaseError は取引先ユースケースのエラーを HTTP レスポンスに変換する。
// エラーを処理した場合 true を返す。
func writeSupplierUsecaseError(w http.ResponseWriter, err error, logPrefix string) bool {
	if err == nil {
		return false
	}

	var vErr *usecase.ValidationError
	if errors.As(err, &vErr) {
		details := make([]ErrorDetail, len(vErr.Violations))
		for i, v := range vErr.Violations {
			details[i] = ErrorDetail{Field: v.Field, Message: v.Message}
		}
		respondWithError(w, http.StatusBadRequest, ValidationErrorCode, "入力内容に誤りがあります", details)
		return true
	}
	if errors.Is(err, domain.ErrSupplierNotFound) {
		respondWithError(w, http.StatusNotFound, NotFoundCode, "取引先が見つかりません", nil)
		return true
	}

	log.Printf("%s: %v", logPrefix, err)
	respondWithError(w, http.StatusInternalServerError, InternalErrorCode, "サーバー内部でエラーが発生しました", nil)
	return true
}
